csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs

```diff
@@ -1607,104 +1641,102 @@ public class ModelMetadata
 
                 // Process version
                 NativeApiStatus.VerifySuccess(NativeMethods.OrtModelMetadataGetVersion(modelMetadataHandle, out _version));
 
 
                 // Process CustomMetadata Map
-                IntPtr customMetadataMapKeysHandle = IntPtr.Zero;
-                long numKeys;
-                NativeApiStatus.VerifySuccess(NativeMethods.OrtModelMetadataGetCustomMetadataMapKeys(modelMetadataHandle, allocator.Pointer, out customMetadataMapKeysHandle, out numKeys));
-
+                NativeApiStatus.VerifySuccess(NativeMethods.OrtModelMetadataGetCustomMetadataMapKeys(modelMetadataHandle,
+                    allocator.Pointer, out IntPtr customMetadataMapKeysHandle, out long numKeys));
                 // We have received an array of null terminated C strings which are the keys that we can use to lookup the custom metadata map
                 // The OrtAllocator will finally free the customMetadataMapKeysHandle
-                using (var ortAllocationKeysArray = new OrtMemoryAllocation(allocator, customMetadataMapKeysHandle, 0))
-                using (var ortAllocationKeys = new DisposableList<OrtMemoryAllocation>((int)numKeys))
+                try
                 {
-                    // Put all the handles to each key in the DisposableList to be disposed off in an exception-safe manner
-                    for (int i = 0; i < (int)numKeys; ++i)
-                    {
-                        ortAllocationKeys.Add(new OrtMemoryAllocation(allocator, Marshal.ReadIntPtr(customMetadataMapKeysHandle, IntPtr.Size * i), 0));
-                    }
-
-                    // Process each key via the stored key handles
-                    foreach (var allocation in ortAllocationKeys)
+                    using (var ortAllocationKeys = new DisposableList<OrtMemoryAllocation>((int)numKeys))
                     {
-                        IntPtr keyHandle = allocation.Pointer;
-                        IntPtr valueHandle = IntPtr.Zero;
-                        NativeApiStatus.VerifySuccess(NativeMethods.OrtModelMetadataLookupCustomMetadataMap(modelMetadataHandle, allocator.Pointer, keyHandle, out valueHandle));
-
-                        using (var ortAllocationValue = new OrtMemoryAllocation(allocator, valueHandle, 0))
+                        // Put all the handles to each key in the DisposableList to be disposed off in an exception-safe manner
+                        for (int i = 0; i < (int)numKeys; ++i)
                         {
-                            var key = NativeOnnxValueHelper.StringFromNativeUtf8(keyHandle);
-                            var value = NativeOnnxValueHelper.StringFromNativeUtf8(valueHandle);
-
-                            // Put the key/value pair into the dictionary
-                            _customMetadataMap[key] = value;
+                            ortAllocationKeys.Add(new OrtMemoryAllocation(allocator, Marshal.ReadIntPtr(customMetadataMapKeysHandle, IntPtr.Size * i), 0));
```

Comment:
Most of the `Marshal.*` APIs are very slow or pessimize some behavior due to back-compat.

In .NET Core, most of the APIs have been "replaced" with other recommended APIs. For example, rather than `Marshal.AllocHGlobal` it is recommended you use `NativeMemory.Alloc`.

In the case of APIs like `Marshal.ReadIntPtr`, it is better to use `System.Runtime.CompilerServices.Unsafe.ReadUnaligned` or to do a simple pointer dereference. Either of these options work on .NET Standard/.NET Framework as well (`Unsafe` is available via a NuGet package, pointers just require you to cast and dereference).